xla/service/gpu/runtime/nccl_collective_permute_thunk.cc

```diff
@@ -114,38 +132,53 @@ NcclCollectivePermuteStartThunk::NcclCollectivePermuteStartThunk(
       .value();
 }
 
+absl::Status NcclCollectivePermuteStartThunk::Initialize(
+    const InitializeParams& params) {
+  TF_RETURN_IF_ERROR(NcclCollectiveThunk::Initialize(params));
+  if (p2p_memcpy_enabled_) {
+    TF_ASSIGN_OR_RETURN(const int64_t current_id,
+                        GetCurrentId(params.collective_params, config_));
+    {
+      absl::MutexLock lock(&mutex_);
+      if (recv_ptr_map_.find(current_id) == recv_ptr_map_.end()) {
+        recv_ptr_map_[current_id] = tsl::MakeUnconstructedAsyncValueRef<void*>();
+      }
+    }
+  }
+
+  return absl::OkStatus();
+}
+
 absl::Status NcclCollectivePermuteStartThunk::RunNcclCollective(
     const ExecuteParams& params, se::Stream& stream,
-    NcclApi::NcclCommHandle comm) {
+    NcclCommHandleWrapper comm_wrapper) {
   TF_ASSIGN_OR_RETURN(
       std::vector<DeviceBufferPair> device_buffers,
       ConvertToDeviceBuffers(params, {buffer_},
                              config_.config.operand_element_type));
   TF_RET_CHECK(device_buffers.size() == 1) << "Expected one buffer pair.";
-
-  GlobalDeviceId global_device_id = params.collective_params->global_device_id;
-
-  TF_ASSIGN_OR_RETURN(const DeviceAssignment::LogicalID current_logical_id,
-                      params.collective_params->device_assn->LogicalIdForDevice(
-                          global_device_id));
-  const int64_t current_id =
-      config_.config.group_mode == CollectiveOpGroupMode::kCrossReplica
-          ? current_logical_id.replica_id
-          : current_logical_id.computation_id;
+  TF_ASSIGN_OR_RETURN(const int64_t current_id,
+                      GetCurrentId(params.collective_params, config_));
   std::string device_string = GetDeviceString(*params.collective_params);
 
   const NcclP2PConfig::SourceTargetMapEntry source_target =
       NcclP2PConfig::GetSourceTarget(config_.id_to_source_target, current_id);
 
-  return ::xla::gpu::RunCollectivePermute(nccl_api(), source_target,
-                                          device_buffers[0], stream, comm,
-                                          device_string, current_id);
+  bool use_memcpy = false;
+  use_memcpy = comm_wrapper.is_local &&
+                    recv_ptr_map_.find(current_id) != recv_ptr_map_.end() &&
+                    p2p_memcpy_enabled_;
+  return ::xla::gpu::RunCollectivePermute(
+      nccl_api(), source_target, device_buffers[0], stream,
+      comm_wrapper.comm_handle, device_string, current_id, use_memcpy,
+      recv_ptr_map_);
 }
 
 absl::Status RunCollectivePermute(
     NcclApi* nccl_api, NcclP2PConfig::SourceTargetMapEntry source_target,
     DeviceBufferPair& buffer, se::Stream& stream, NcclApi::NcclCommHandle comm,
-    absl::string_view device_string, int64_t current_id) {
+    absl::string_view device_string, int64_t current_id, bool use_memcpy,
+    absl::node_hash_map<int64_t, AsyncValueRef<void*>>& recv_ptr_map) {
```

Comment:
If you'll pass map by const-ref then I think it should correct wrt to data races, you just have to promise that all insertions into the map a properly guarded

Reply:
yup, updated.